### PR TITLE
fix parse_version, add tests, fixes #2556

### LINKED ISSUE #2556: version parsing code blows up with a dirty workdir

### PR DESCRIPTION
**Pull request by ThomasWaldmann**

### CODE REVIEW

File: src/borg/version.py [hunk 1]
```diff
@@ -17,19 +17,18 @@ def parse_version(version):
 
     This version format is part of the remote protocol, don‘t change in breaking ways.
     """
-
-    parts = version.split('+')[0].split('.')
```
Comment (ThomasWaldmann) [inline src/borg/version.py]: splitting on "+" lead to issues on setuptools < 8.0.

File: src/borg/version.py [hunk 2]
```diff
@@ -17,19 +17,18 @@ def parse_version(version):
...
-    if parts[-1].startswith('dev'):
-        del parts[-1]
```
Comment (ThomasWaldmann) [inline src/borg/version.py]: this did not work when there was a "-" instead of a "+" in the version string.

File: src/borg/testsuite/version.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+from ..version import parse_version
+
+
+def test_parse_version():
+    # the version parsed ignores any part after "dev", these tests are just to
+    # check it can cope with such versions.
+    # setuptools < 8.0 uses "devXXX-"
+    assert parse_version('1.0.0a1.dev204-g8866961.d20170606') == (1, 0, 0, -4, 1)
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/version.py]: versions like this ^^^ blew up the old code.

File: src/borg/testsuite/version.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+from ..version import parse_version
+
+
+def test_parse_version():
```
Comment (RonnyPfannschmidt) [inline src/borg/testsuite/version.py]: this one begs for a parameterize and a dictionary

**Comment by ThomasWaldmann** [comment]
@RonnyPfannschmidt ok, parametrize is there. not sure what you wanted with the dict?

**Comment by textshell** [comment]
@RonnyPfannschmidt is there any upstream documentation on the various formats used in different setup tools versions? It seems that would be very useful to have a complete parser without surprises (not just for borg)

**Comment by ThomasWaldmann** [comment]
@textshell and if it is `1.1.0+dYYYMMMDD` now, was it `1.1.0-dYYYMMMDD` before setuptools 8.0?

**Comment by textshell** [comment]
I have no clue. Maybe it can't even generate a version like this. I tried looking in the implementation but really don't grok what is going on.
I guess we can take wild guesses, just chop off everything starting with the first `+` or try to get some more information.

I think i originally tried just going with choping off everything starting with `+`. Maybe just add that back to the code that new also understands other but disjunkt special cases?

Or maybe split on `.`, `+` and `-` and discard from the first item that does not match `([0-9]+)(a|b|rc)?([0-9]+)`?

**Comment by enkore** [comment]
Why are we using this automation? It does not seem to save time or troubles.

**Comment by textshell** [comment]
i don't do any borg release work. But i would not mind if borg would just have the version somewhere in a file tracked in git.
(Besides, looking at a random git commit in github and having some way to see what version span that commit came from is harder if the version is nowhere in the files, borg of course has a changelog that usually works to find the version)

One nice thing about a scm based version is that in theory it allows to see where something that is not a real version is roughly on the development time line. i.e. for debug output.

**Comment by ThomasWaldmann** [comment]
i like that the version is derived from git. that way, it is immediately obvious which version people run (not just roughly as the case if we would update manually now and then).

**Comment by ThomasWaldmann** [comment]
I switched to a different approach:
use a regex to start parsing from the left, only parse as long as it matches stuff we actually want.
and do not support x.y version numbers any more, we do not use them anyway and it is much simpler that way.

**Comment by textshell** [comment]
I wonder if we want to terminate the regex with something that constrains the following chars. currently a version like 1.1.0abc would parse as 1.1.0. But greediness currently ensures that all digits are picked up. So i'm not sure if something like `$|[^a-zA-Z0-9] would be of any practical use at the end of the regex.

Explicitly restricting the parser to major.minor.patch seems like a good simplification. Much less guessing. We just need to make sure not to have versions that don't fit without adjusting that code. Preferably by not having version that do not fit into that form.

**Comment by ThomasWaldmann** [comment]
well, practically we only have to parse what setuptools_scm would generate (or what we would put into some constant in the src).

**Comment by textshell** [comment]
indeed. We only need to prove against anything setuptools_scm did produce or might produce in the future.
I think this should work in the current form.

**Pull request merged**